torch/lib/THD/master_worker/master/generic/THDTensor.cpp

```diff
@@ -635,9 +760,134 @@ void THDTensor_(unfold)(THDTensor *self, THDTensor *src,
+    if (src->size[d] != 1) {
+      if (d != ndim) {
+        self->size[ndim] = src->size[d];
+        self->stride[ndim] = src->stride[d];
+      }
+      ndim++;
+    }
+  }
+
+  /* right now, we do not handle 0-dimension tensors */
+  if(ndim == 0 && src->nDimension > 0) {
+    self->size[0] = 1;
+    self->stride[0] = 1;
+    ndim = 1;
+  }
+  self->nDimension = ndim;
+  masterCommandChannel->sendMessage(
+      packMessage(Functions::tensorSqueeze, self, src),
+      THDState::s_current_worker
+  );
+}
+
+void THDTensor_(squeeze1d)(THDTensor *self, THDTensor *src, int dimension) {
+  if(!src)
+    src = self;
+
+  THArgCheck((dimension >= 0) && (dimension < src->nDimension), 2, "dimension out of range");
+
+  THDTensor_(set)(self, src);
+
+  if (src->size[dimension] == 1 && src->nDimension > 1) {
+    for (std::size_t d = dimension; d < self->nDimension-1; d++) {
+      self->size[d] = self->size[d+1];
+      self->stride[d] = self->stride[d+1];
+    }
+    self->nDimension--;
+  }
+  masterCommandChannel->sendMessage(
+      packMessage(Functions::tensorSqueeze1d, self, src),
+      THDState::s_current_worker
+  );
+}
+
+int THDTensor_(isContiguous)(const THDTensor *self) {
+  long z = 1;
+  for (std::size_t d = self->nDimension - 1; d >= 0; d--) {
+    if (self->size[d] != 1) {
+      if (self->stride[d] == z)
+        z *= self->size[d];
+      else
+        return 0;
+    }
+  }
+  return 1;
+}
+
 int THDTensor_(isSameSizeAs)(const THDTensor *self, const THDTensor *src) {
-  throw std::runtime_error("isSameSizeAs not implemented yet");
+  if (self->nDimension != src->nDimension)
+    return 0;
+  for (std::size_t d = 0; d < self->nDimension; d++)
+    if (self->size[d] != src->size[d])
+      return 0;
+  return 1;
+}
+
+int THDTensor_(isSetTo)(const THDTensor *self, const THDTensor *src) {
+  if (!self->storage)
+    return 0;
+  if (self->storage == src->storage &&
+      self->storageOffset == src->storageOffset &&
+      self->nDimension == src->nDimension) {
+    for (std::size_t d = 0; d < self->nDimension; d++) {
+      if (self->size[d] != src->size[d] || self->stride[d] != src->stride[d])
+        return 0;
+    }
+    return 1;
+  }
+  return 0;
+}
+
+int THDTensor_(isSize)(const THDTensor *self, const THLongStorage *dims) {
+  if (self->nDimension != dims->size)
+    return 0;
+  for (std::size_t d = 0; d < self->nDimension; d++)
+    if (self->size[d] != dims->data[d])
+      return 0;
+  return 1;
+}
+
+ptrdiff_t THDTensor_(nElement)(const THDTensor *self) {
+  if (self->nDimension == 0) {
+    return 0;
+  } else {
+    ptrdiff_t nElement = 1;
+    for (std::size_t d = 0; d < self->nDimension; d++) {
+      nElement *= self->size[d];
+    }
+    return nElement;
+  }
+}
+
+void THDTensor_(retain)(THDTensor *tensor) {
+  if (tensor->flag & TH_TENSOR_REFCOUNTED)
+    THAtomicIncrementRef(&tensor->refcount);
+}
+
+void THDTensor_(free)(THDTensor *tensor) {
+  // TODO: free storage?
+  if (!(--tensor->refcount)) {
```

Comment:
Shouldn't this be some kind of `THAtomicDecrementRef`?

Btw. You do not do the flag checking `if(self->flag & TH_TENSOR_REFCOUNTED)` or `if (!tensor) return`.